products/kms/api.yaml

```diff
@@ -56,3 +56,74 @@ objects:
         'Creating a key ring':
           'https://cloud.google.com/kms/docs/creating-keys#create_a_key_ring'
       api: 'https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings'
+  - !ruby/object:Api::Resource
+    name: 'CryptoKey'
+    base_url: '{{key_ring}}/cryptoKeys?cryptoKeyId={{name}}'
+    self_link: '{{key_ring}}/cryptoKeys/{{name}}'
+    update_verb: :PATCH
+    update_mask: true
+    description: |
+      A `CryptoKey` represents a logical key that can be used for cryptographic operations.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          The resource name for the CryptoKey.
+        required: true
+        input: true
+      - !ruby/object:Api::Type::Time
+        name: 'creationTime'
+        description: |
+          The time that this resource was created on the server.
+          This is in RFC3339 text format.
+        output: true
+      - !ruby/object:Api::Type::String
+        name: 'keyRing'
+        description: |
+          The id of the `KeyRing` to which the key shall belong.
+        required: true
+        input: true
+      - !ruby/object:Api::Type::Enum
+        name: 'purpose'
+        description: |
+          The immutable purpose of this CryptoKey. See the
+          [purpose reference](https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings.cryptoKeys#CryptoKeyPurpose)
+          for possible inputs.
+        values:
+          - "ENCRYPT_DECRYPT"
+          - "ASYMMETRIC_SIGN"
+          - "ASYMMETRIC_DECRYPT"
+        default_value: :ENCRYPT_DECRYPT
+        input: true
+      - !ruby/object:Api::Type::String
+        name: 'rotationPeriod'
+        description: |
+          Every time this period passes, generate a new CryptoKeyVersion and set it as the primary.
+          The first rotation will take place after the specified period. The rotation period has
+          the format of a decimal number with up to 9 fractional digits, followed by the
+          letter `s` (seconds). It must be greater than a day (ie, 86400).
+      - !ruby/object:Api::Type::NestedObject
+        name: 'versionTemplate'
+        description: |
+          A template describing settings for new crypto key versions.
+        properties:
+          - !ruby/object:Api::Type::String
+            name: 'algorithm'
+            description: |
+               The algorithm to use when creating a version based on this template.
+               See the [algorithm reference](https://cloud.google.com/kms/docs/reference/rest/v1/CryptoKeyVersionAlgorithm) for possible inputs.
+            required: true
+            input: true
```

Comment:
This was previously not ForceNew- was that a bug, or is this an oversight?

Reply:
It was effectively a ForceNew before since `version_template` has only `algorithm` and `protection_level`,  `protection_level` is never updatable (so ForceNew), and `algorithm` can only be `GOOGLE_SYMMETRIC_ENCRYPTION` for symmetric keys. But with asymmetric keys which we will support now with `purpose` we should be able to update `algorithm`, you're right. I'll fix this.

Reply:
Interesting, got hit by the old `updateMask` limitations which my custom `update_mask_fields` fortunately also works around:
```
  # google_kms_crypto_key.my-key-2 will be updated in-place
  ~ resource "google_kms_crypto_key" "my-key-2" {
        id        = "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2"
        key_ring  = "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring"
        name      = "crypto-key-2"
        purpose   = "ASYMMETRIC_DECRYPT"
        self_link = "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2"

        timeouts {}

      ~ version_template {
          ~ algorithm        = "RSA_DECRYPT_OAEP_3072_SHA256" -> "RSA_DECRYPT_OAEP_4096_SHA256"
            protection_level = "HSM"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
standard PATCH fails:
```
PATCH /v1/projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2?alt=json&updateMask=versionTemplate HTTP/1.1
Host: cloudkms.googleapis.com
User-Agent: Terraform/0.12.0 (+https://www.terraform.io) terraform-provider-google/dev
Content-Length: 89
Content-Type: application/json
Accept-Encoding: gzip

{
 "versionTemplate": {
  "algorithm": "RSA_DECRYPT_OAEP_4096_SHA256",
  "protectionLevel": "HSM"
 }
}

-----------------------------------------------------
2019/06/12 23:17:53 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 400 Bad Request
Alt-Svc: quic=":443"; ma=2592000; v="46,44,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 12 Jun 2019 21:17:53 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "error": {
    "code": 400,
    "message": "Field crypto_key.version_template.protection_level may not be updated. The value is immutable.",
    "status": "INVALID_ARGUMENT"
  }
}

-----------------------------------------------------
```
but succeeds with a custom `update_mask_fields`:
```
PATCH /v1/projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2?alt=json&updateMask=versionTemplate.algorithm HTTP/1.1
Host: cloudkms.googleapis.com
User-Agent: Terraform/0.12.0 (+https://www.terraform.io) terraform-provider-google/dev
Content-Length: 89
Content-Type: application/json
Accept-Encoding: gzip

{
 "versionTemplate": {
  "algorithm": "RSA_DECRYPT_OAEP_4096_SHA256",
  "protectionLevel": "HSM"
 }
}

-----------------------------------------------------
2019/06/12 23:35:57 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="46,44,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 12 Jun 2019 21:35:57 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2",
  "purpose": "ASYMMETRIC_DECRYPT",
  "createTime": "2019-06-02T14:15:09.913833168Z",
  "versionTemplate": {
    "protectionLevel": "HSM",
    "algorithm": "RSA_DECRYPT_OAEP_4096_SHA256"
  }
}

-----------------------------------------------------
```
Pushed my changes now.